jetty-util/src/main/java/org/eclipse/jetty/util/MutableIndex.java

```diff
@@ -0,0 +1,79 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2020 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under
+// the terms of the Eclipse Public License 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0
+//
+// This Source Code may also be made available under the following
+// Secondary Licenses when the conditions for such availability set
+// forth in the Eclipse Public License, v. 2.0 are satisfied:
+// the Apache License v2.0 which is available at
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.util;
+
+import java.util.Map;
+
+public interface MutableIndex<V> extends Index<V>
+{
+    boolean isFull();
+
+    /**
+     * Put an entry into the index.
+     *
+     * @param s The key for the entry
+     * @param v The value of the entry
+     * @return True if the index had capacity to add the field.
+     */
+    boolean put(String s, V v);
+
+    /**
+     * Put a value as both a key and a value.
+     *
+     * @param v The value and key
+     * @return True if the Trie had capacity to add the field.
+     */
+    boolean put(V v);
+
+    void clear();
+
+    V remove(String s);
+
+    class Builder<V> extends Index.Builder<V>
+    {
+        private int fixedCapacity;
+
+        Builder(Map<String, V> contents, boolean caseSensitive)
+        {
+            super(contents, caseSensitive);
+        }
+
+        public MutableIndex.Builder<V> fixedCapacity(int capacity)
```

Comment:
We are going to need good javadoc to explain what units `capacity` is in!

perhaps also have a `fixedCapacity(double)` that multiplies the capacity of what is required for anything added to the builder.

Reply:
Also for a mutable Index builder, this would be better named `maxCapacity`... which might be implemented as a fixedCapacity at max, or might not.

Reply:
I like the `maxCapacity` name better, so I'll rename it.I kinda like the idea of using `maxCapacity` as a growth factor, except that the `HttpParser`'s `_fieldCache` has a capacity that is user-configurable, and I don't think the intent is to use a growing index.